File: works-text/src/main/scala/de/kp/works/text/topic/LDAPredictor.scala
package de.kp.works.text.topic
/*
 * Copyright (c) 2019 Dr. Krusche & Partner PartG. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @author Stefan Krusche, Dr. Krusche & Partner PartG
 * 
 */

import java.util.{Map => JMap}

import org.apache.spark.ml.clustering.LDAModel
import org.apache.spark.ml.linalg.Vector

import org.apache.spark.sql._
import org.apache.spark.sql.functions._

import de.kp.works.text.AnnotationBase
import de.kp.works.text.embeddings.Word2VecModel

/*
 * In contrast to say, KMeans, where each entity can only belong to a single cluster
 * (hrad-clustering), LDA allows to interpret the topic probabilities as the proportion
 * of cluster membership (soft-clustering), or 'fuzzy clustering'.
 * 
 * This provides a more nuanced way of recommending similar items, finding duplicates,
 * or discovering user profiles/personas.
 * 
 * LDALabeler supports use cases where hard-clustering based on LDA is required.
 */
class LDALabeler(model:LDAModel, word2vec:Word2VecModel) extends AnnotationBase {
  
  def label(dataset:Dataset[Row], textCol:String, labelCol:String, params:JMap[String,Object]):Dataset[Row] = {
    /*
     * STEP #1: Leverage the LDA predictor to determine
     * the topic distribution
     */
    val predictions = new LDAPredictor(model, word2vec).predict(dataset, textCol, "topics", params)
    /*
     * Determine the index of the topic with the highest
     * contribution to the document
     */
    val topics2label = udf{topics:Vector => {
      
      val zipped = topics.toArray.zipWithIndex
      val label = zipped.sortBy(- _._1).head._1
      
      label.toDouble
    }}

    val output = predictions.withColumn(labelCol, topics2label(col("topics")))
    output.drop("topics")

  }
  
}

/*
 * The LDAPredictor is used to assign a certain label distribution (can be intepreted as
 * topics) to a text document. The label distribution is computed by leveraging a pre-
 * trained LDA model based on doument embeddings 
 */
class LDAPredictor(model:LDAModel, word2vec:Word2VecModel) extends AnnotationBase {
   
  def predict(dataset:Dataset[Row], textCol:String, topicCol:String, params:JMap[String,Object]):Dataset[Row] = {
    /*
     * STEP #1: Extracted normalized tokens and transform
     * each token into an embedding vector
     */
    var document = normalizedTokens(dataset, textCol)

    word2vec.setInputCols(Array("sentences", "token"))
    word2vec.setOutputCol("embeddings")
    
    document = word2vec.transform(document)
    /*
     * STEP #2: Transform token embeddings into a document
     * feature vector
     */
    val strategy = if (params.containsKey("strategy")) params.get("strategy").asInstanceOf[String] else "AVERAGE"
    val embeddings2vector = embeddings2vector_udf(strategy)
    
    val dropCols = Array("document", "sentences", "token", "embeddings")
    document = document.withColumn("vector", embeddings2vector(col("embeddings"))).drop(dropCols: _*)
    
    model.setFeaturesCol("vector")
    /*
     * The model adds an internal column "topicDistribution" 
     * to the document; this is a k-dimensional vector that
     * determines the portion each topic contributes to the
     * document (line)  
     */
    document = model.transform(document)
      .withColumnRenamed("topicDistribution", topicCol).drop("vector")

    document
    
  }
  
  def devectorize(dataset:Dataset[Row], vectorCol:String, featureCol:String): Dataset[Row] = {
    /*
     * The dataset contains an Apache Spark Vector and must be transformed
     * into an Array of Double value (to CDAP Structured Record)
     */
    val devector_udf = udf {vector:Vector => {
      vector.toArray
    }}

    dataset.withColumn(featureCol, devector_udf(col(vectorCol)))
    
  }
 
}

File: works-ml/src/main/java/de/kp/works/ml/prediction/LDAPredictor.java
package de.kp.works.ml.prediction;
/*
 * Copyright (c) 2019 - 2021 Dr. Krusche & Partner PartG. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @author Stefan Krusche, Dr. Krusche & Partner PartG
 * 
 */

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.ml.clustering.LDAModel;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

import io.cdap.cdap.api.annotation.Description;
import io.cdap.cdap.api.annotation.Name;
import io.cdap.cdap.api.annotation.Plugin;
import io.cdap.cdap.api.data.schema.Schema;
import io.cdap.cdap.etl.api.PipelineConfigurer;
import io.cdap.cdap.etl.api.StageConfigurer;
import io.cdap.cdap.etl.api.batch.SparkCompute;
import io.cdap.cdap.etl.api.batch.SparkExecutionPluginContext;

import de.kp.works.core.predictor.PredictorCompute;
import de.kp.works.core.recording.clustering.LDARecorder;
import de.kp.works.core.cluster.PredictorConfig;
import de.kp.works.core.recording.MLUtils;

@Plugin(type = SparkCompute.PLUGIN_TYPE)
@Name("LDAPredictor")
@Description("A prediction stage that leverages a trained Apache Spark ML Latent Dirichlet Allocation (LDA) clustering model.")
public class LDAPredictor extends PredictorCompute {

	private static final long serialVersionUID = 1979301843408604941L;

	private final PredictorConfig config;
	private LDAModel model;

	public LDAPredictor(PredictorConfig config) {
		this.config = config;
	}

	@Override
	public void initialize(SparkExecutionPluginContext context) throws Exception {
		config.validate();

		LDARecorder recorder = new LDARecorder(configReader);
		/* 
		 * STEP #1: Retrieve the trained clustering model
		 * that refers to the provide name, stage and option
		 */
		model = recorder.read(context, config.modelName, config.modelStage, config.modelOption);
		if (model == null)
			throw new IllegalArgumentException(String.format("[%s] A clustering model with name '%s' does not exist.",
					this.getClass().getName(), config.modelName));

		/* 
		 * STEP #2: Retrieve the profile of the trained
		 * clustering model for subsequent annotation
		 */
		profile = recorder.getProfile();

	}

	@Override
	public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {

		config.validate();

		StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
		/*
		 * Try to determine input and output schema; if these schemas are not explicitly
		 * specified, they will be inferred from the provided data records
		 */
		inputSchema = stageConfigurer.getInputSchema();
		if (inputSchema != null) {
			validateSchema(inputSchema);
			/*
			 * In cases where the input schema is explicitly provided, we determine the
			 * output schema by explicitly adding the prediction column
			 */
			outputSchema = getOutputSchema(inputSchema, config.predictionCol);
			stageConfigurer.setOutputSchema(outputSchema);

		}

	}

	/**
	 * A helper method to compute the output schema in that use cases where an input
	 * schema is explicitly given
	 */
	@Override
	public Schema getOutputSchema(Schema inputSchema, String predictionField) {

		assert inputSchema.getFields() != null;
		List<Schema.Field> fields = new ArrayList<>(inputSchema.getFields());
		
		fields.add(Schema.Field.of(predictionField, Schema.arrayOf(Schema.of(Schema.Type.DOUBLE))));
		return Schema.recordOf(inputSchema.getRecordName() + ".predicted", fields);

	}	
	
	/**
	 * This method computes predictions either by applying a trained LDA
	 * clustering model; as a result, the source dataset is enriched by an
	 * extra column (predictionCol) that specifies the target variable in 
	 * form of an Array[Double] (which specifies the topic distribution)
	 */
	@Override
	public Dataset<Row> compute(SparkExecutionPluginContext context, Dataset<Row> source) throws Exception {
		/*
		 * STEP #1: Extract configuration parameters
		 */
		String featuresCol = config.featuresCol;
		String predictionCol = config.predictionCol;
		/*
		 * The vectorCol specifies the internal column that has to be built from the
		 * featuresCol and that is used for prediction purposes
		 */
		String vectorCol = "_vector";
		/*
		 * Prepare provided dataset by vectorizing the feature column which is specified
		 * as Array[Numeric]
		 */
		Dataset<Row> vectorset = MLUtils.vectorize(source, featuresCol, vectorCol, true);

		model.setFeaturesCol(vectorCol);
		/*
		 * The output column with estimates of the topic mixture distribution for each document
		 * (often called "theta" in the literature) is set internally to 'topicDistribution'
		 */
		Dataset<Row> predictions = MLUtils.devectorize(
				model.transform(vectorset), "topicDistribution", predictionCol);
		/*
		 * Remove intermediate vector column from predictions
		 * and annotate each prediction with the model profile
		 */
		Dataset<Row> output = predictions.drop(vectorCol);
		return annotate(output, CLUSTER_TYPE);

	}

	@Override
	public void validateSchema(Schema inputSchema) {
		config.validateSchema(inputSchema);
	}

}
